plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -57,62 +58,99 @@ import           PlutusTx.Prelude                 hiding (Semigroup (..), unless
 import           PlutusTx.Sqrt
 import           Prelude                          (Semigroup (..))
 import qualified Prelude
-import           Text.Printf                      (printf)
+import           Text.Printf                      (PrintfArg, printf)
 
 uniswapTokenName, poolStateTokenName :: TokenName
 uniswapTokenName = "Uniswap"
 poolStateTokenName = "Pool State"
 
--- | A handy alias to put things in the language of "Coins" instead of
--- "AssetClass".
-type Coin = AssetClass
-
 -- Note: An orphan instance here because of the alias above.
 deriving anyclass instance ToSchema AssetClass
 
+-- | Uniswap coin token
+data U = U
+
+-- | "A"-side coin token
+data A = A
+
+-- | "B"-side coin token
+data B = B
+
+-- | Pool-state coin token
+data PoolState = PoolState
+
+-- | Liquidity-state coin token
+data Liquidity = Liquidity
+
+-- | A single 'AssetClass'. Because we use three coins, we use a phantom type to track
+-- which one is which.
+newtype Coin a = Coin { unCoin :: AssetClass }
+  deriving stock   (Show, Generic)
+  deriving newtype (ToJSON, FromJSON, ToSchema, Eq, Prelude.Eq, Prelude.Ord)
+
+-- | Likewise for 'Integer'; the corresponding amount we have of the
+-- particular 'Coin'.
+newtype Amount a = Amount { unAmount :: Integer }
```

Comment:
I wonder if these are generally useful enough to put in a library? Not much to them, I guess.

Reply:
It's possible; maybe we can just keep an eye out if any other examples would benefit from it; I have a sense that probably they will; but it might be combined with some other refactoring of the AssetClass stuff.